go_lib/deckhouse-config/validator.go

```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
+
+	ValidateCRWarning string
+	ValidateCRError   string
+	ConversionError   string
+	ValidationError   string
+}
+
+func (v ValidationResult) HasError() bool {
+	return v.ConversionError != "" || v.ValidateCRError != "" || v.ValidationError != ""
 }
 
-// Validate checks version value if spec.settings value is not empty,
-// then converts spec.settings to the latest version and finally validates
-// spec.settings using OpenAPI schema defined in related config-values.yaml file.
-func (c *ConfigValidator) Validate(cfg *d8cfg_v1alpha1.ModuleConfig) (ValidationResult, error) {
+func (v ValidationResult) Error() string {
+	errs := make([]string, 0)
+	if v.ValidateCRError != "" {
+		errs = append(errs, v.ValidateCRError)
+	}
+	if v.ConversionError != "" {
+		errs = append(errs, v.ConversionError)
+	}
+	if v.ValidationError != "" {
+		errs = append(errs, v.ValidationError)
+	}
+	return strings.Join(errs, ", ")
+}
+
+// ValidateCR checks if ModuleConfig resource is well-formed.
+func (c *ConfigValidator) ValidateCR(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationResult {
 	result := ValidationResult{}
 
-	// It is ok if resource spec is empty or if it has enabled field only.
-	if len(cfg.Spec.Settings) == 0 && cfg.Spec.Version == 0 {
-		return result, nil
+	if cfg.Spec.Version == 0 {
+		// Resource is not valid when spec.settings are specified without version.
+		if cfg.Spec.Settings != nil {
+			result.ValidateCRError = "spec.version is required when spec.settings are specified"
+		}
+		// Resource is valid without spec.version and spec.settings.
+		return result
 	}
-	if cfg.Spec.Settings == nil && cfg.Spec.Version > 0 {
-		return result, fmt.Errorf("spec.version is forbidden when settings are not specified")
+
+	// Can run conversions and validations if spec.version and spec.settings are specified.
+	if cfg.Spec.Settings == nil {
+		// Warn about spec.version without spec.settings.
+		result.ValidateCRWarning = "spec.version is redundant without spec.settings"
+	} else {
+		// Resource has both spec.settings and spec.version — it is ok to convert and validate with OpenAPI schema.
+		result.HasVersionedSettings = true
 	}
 
-	// Settings are present, check version using conversion chain.
+	// Check if there is registered conversion for the version and if the version is the latest.
 	chain := conversion.Registry().Chain(cfg.GetName())
 	latestVer := chain.LatestVersion()
 
+	// Check if version is unknown.
 	if !chain.IsKnownVersion(cfg.Spec.Version) {
 		previousVersions := concatIntList(chain.PreviousVersionsList())
 		if previousVersions != "" {
 			previousVersions = fmt.Sprintf(", or one of previous versions: %s", previousVersions)
 		}
 
-		return result, fmt.Errorf("spec.version=%d is invalid. Use latest version %d%s", cfg.Spec.Version, latestVer, previousVersions)
+		msg := fmt.Sprintf("spec.version=%d is unsupported. Use latest version %d%s", cfg.Spec.Version, latestVer, previousVersions)
+		if result.HasVersionedSettings {
```

Comment:
On line 99 you just set effectively 
```go
result.HasVersionedSettings = cfg.Spec.Settings != nil
```

And here you check the same fact.

Reply:
In general I'd not distinguish empty settings from any other state. Why would we want to give it special attention?

Reply:
I've removed this public field, it is now a helper func.